arbnode/blockmetadata.go

```diff
@@ -0,0 +1,167 @@
+package arbnode
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/spf13/pflag"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/execution"
+	"github.com/offchainlabs/nitro/execution/gethexec"
+	"github.com/offchainlabs/nitro/util/signature"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+type BlockMetadataRebuilderConfig struct {
+	Enable          bool          `koanf:"enable"`
+	Url             string        `koanf:"url"`
+	JWTSecret       string        `koanf:"jwt-secret"`
+	RebuildInterval time.Duration `koanf:"rebuild-interval"`
+	APIBlocksLimit  uint64        `koanf:"api-blocks-limit"`
+}
+
+var DefaultBlockMetadataRebuilderConfig = BlockMetadataRebuilderConfig{
+	Enable:          false,
+	RebuildInterval: time.Minute * 5,
+	APIBlocksLimit:  100,
+}
+
+func BlockMetadataRebuilderConfigAddOptions(prefix string, f *pflag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultBlockMetadataRebuilderConfig.Enable, "enable syncing blockMetadata using a bulk metadata api")
+	f.String(prefix+".url", DefaultBlockMetadataRebuilderConfig.Url, "url for bulk blockMetadata api")
+	f.String(prefix+".jwt-secret", DefaultBlockMetadataRebuilderConfig.JWTSecret, "filepath of jwt secret")
+	f.Duration(prefix+".rebuild-interval", DefaultBlockMetadataRebuilderConfig.RebuildInterval, "interval at which blockMetadata is synced regularly")
+	f.Uint64(prefix+".api-blocks-limit", DefaultBlockMetadataRebuilderConfig.APIBlocksLimit, "maximum number of blocks allowed to be queried for blockMetadata per arb_getRawBlockMetadata query.\n"+
+		"This should be set lesser than or equal to the value set on the api provider side")
+}
+
+type BlockMetadataRebuilder struct {
+	stopwaiter.StopWaiter
+	config BlockMetadataRebuilderConfig
+	db     ethdb.Database
+	client *rpc.Client
+	exec   execution.ExecutionClient
+}
+
+func NewBlockMetadataRebuilder(ctx context.Context, c BlockMetadataRebuilderConfig, db ethdb.Database, exec execution.ExecutionClient) (*BlockMetadataRebuilder, error) {
+	var err error
+	var jwt *common.Hash
+	if c.JWTSecret != "" {
+		jwt, err = signature.LoadSigningKey(c.JWTSecret)
+		if err != nil {
+			return nil, fmt.Errorf("BlockMetadataRebuilder: error loading jwt secret: %w", err)
+		}
+	}
+	var client *rpc.Client
+	if jwt == nil {
+		client, err = rpc.DialOptions(ctx, c.Url)
```

Comment:
the default `wsMessageSizeLimit` is 32MB, it probably will be usually enough***, but to mitigate future issues we can:
* either add a comment here that we don't change the default
* or add a config and set the size limit like here: https://github.com/OffchainLabs/nitro/blob/a3ad260fca1254f9b4c12b597af46237f58b9273/util/rpcclient/rpcclient.go#L271-L275

*** if I get it right, 32MB = 256 Mbit and we should fit bitset metadata for almost 256M txes in a message; with TPS=100, we'd have 25 txes in block, so that would be enough for fetching metadata for around 10M blocks. That seems to be well above what we need, especially when limiting maximal range size on both sides to ~100 blocks. My only worry is that if we put more data into per block metadata, or if an Orbit hits the limit.

Reply:
btw. would it make sense to use `nitro/util/rpclient.RpcClient` here and have `nitro/util/rpclient.ClientConfig` as field of `BlockMetadataRebuilderConfig`?

Reply:
I think its a good idea to use `nitro/util/rpclient.RpcClient`, thanks for the suggestion!